### DISCUSSION TITLE
Wrong form of CrystalPlasticityTwinningKalidindiUpdate Model realized in MOOSE?

### DISCUSSION THREAD

**Discussion by SuperkakaSCU** [idaholab/moose#26831] [Q&A Modules: Solid mechanics]
Hello MOOSE experts,

For the "CrystalPlasticityTwinningKalidindiUpdate.C" Module, I compared the constitutive equations shown on the MOOSE website and the program. I thought there might be some mismatches, and ask for help:

**_Question1: The rate of twin volume fraction on each twin system_**
**Question:**
For "_slip_increment" in "calculateConstitutiveSlipDerivative", it has already considered time step. So, do we still need to contain "_substep_dt" in the rate of twin volume fraction again?

Code:
void CrystalPlasticityTwinningKalidindiUpdate::calculateStateVariableEvolutionRateComponent()
{
for (const auto i : make_range(_number_slip_systems))
_twin_volume_fraction_increment[_qp][i] =
_slip_increment[_qp][i] / _characteristic_twin_shear * **_substep_dt**;
}

**_Question2: Twin Resistance_**
**Question:**
The constitutive equation shown on the website uses the twin fraction increment to calculate the increment of twin propagation. However, in the code, it seems like it uses twin fraction (_twin_volume_fraction[_qp][j]) rather than twin fraction increment.

Code:
void CrystalPlasticityTwinningKalidindiUpdate::calculateTwinResistance()
{
DenseVector twin_hardening_increment(_number_slip_systems);

for (const auto i : make_range(_number_slip_systems))
{
twin_hardening_increment(i) = 0.0;
for (const auto [j](https://github.com/idaholab/moose/issues/0) : make_range(_number_sl[i](https://github.com/idaholab/moose/issues/0)p_systems))
{
if (MooseUtils::relativeFuzzyEqual(_slip_plane_normal[j](https://github.com/idaholab/moose/issues/1), _sl[i](https://github.com/idaholab/moose/issues/1)p_plane_normali) &&
MooseUtils::relativeFuzzyEqual(_slip_plane_normalj, _slip_plane_normali))
// If the first two are the same, the third index will have to be as well
{
if (_slip_increment[_qp][j] > 0.0)
twin_hardening_increment(i) += _coplanar_coefficient_twin_hardening *
_total_twin_volume_fraction[_qp] *
**_twin_volume_fraction[_qp][j]**;
}
else // assume non-coplanar
{
if (_slip_increment[_qp][j] > 0.0)
twin_hardening_increment(i) +=
_non_coplanar_coefficient_twin_hardening *
std::pow(_total_twin_volume_fraction[_qp], _noncoplanar_exponent) *
**_twin_volume_fraction[_qp][j]**;
}
}
.....
}

Thanks for your time!

**Comment by GiudGiud** [comment]
@sapitts

**Reply by jiangwen84** [reply]
I took a quick look at both codes and paper. I felt @SuperkakaSCU is probably correct on those two issues. However, I would like to hear @sapitts 's suggestions before making any code fix.
